File: 934-bitwise-ors-of-subarrays/bitwise-ors-of-subarrays.java
class Solution {
    public int subarrayBitwiseORs(int[] arr) {
        Set<Integer> set = new HashSet<>();

        for (int i = 0; i < arr.length; i++) {
            set.add(arr[i]); // Each single element is a valid subarray

            // Traverse backward from i-1 to 0 to form subarrays ending at i
            for (int j = i - 1; j >= 0; j--) {
                // Optimization: if OR with arr[i] doesn't change arr[j], break early
                if ((arr[i] | arr[j]) == arr[j])
                    break;

                arr[j] |= arr[i]; // Update arr[j] to OR result of arr[j..i]
                set.add(arr[j]); // Add this result to the set
            }
        }

        return set.size();
    }
}

File: 934-bitwise-ors-of-subarrays/bitwise-ors-of-subarrays.go
func subarrayBitwiseORs(arr []int) int {
	set := make(map[int]struct{}) // To store distinct OR values

	for i := 0; i < len(arr); i++ {
		set[arr[i]] = struct{}{} // Add the element itself

		// Go backward to update OR results for subarrays ending at i
		for j := i - 1; j >= 0; j-- {
			if (arr[j] | arr[i]) == arr[j] {
				break // Early exit if no change
			}
			arr[j] |= arr[i]
			set[arr[j]] = struct{}{}
		}
	}

	return len(set)
}
